samcli/lib/build/app_builder.py

```diff
@@ -174,10 +175,13 @@ def _build_function(self, function_name, codeuri, runtime):
         # Code is always relative to the given base directory.
         code_dir = str(pathlib.Path(self._base_dir, codeuri).resolve())
 
+        if code_dir in self._package_cache:
+            return self._package_cache[code_dir]
```

Comment:
Ideally there shouldn't be a build neccessary if the contents have not changed. We would still need a way to conduct a force build

Reply:
Not necessarily true in all cases.  For example, I am building and deploying SAM apps in a CI/CD pipeline.  My repository contains the SAM app as well as unit tests and integration tests.  If I update my tests, it will force a whole build and deploy of the SAM app.  By using this hash, my CI/CD system will still generate the correct artifacts (packaged.yaml referencing the same package in S3) without actually updating the application.

These changes also save a lot of time when my application does change.  Currently, my application has several Lambda functions that all share the same CodeUri.  `sam build` ends up making a copy of my CodeUri folder for each function and `sam package` zips and uploads a copy of my CodeUri folder for each function, which takes more time and generates many duplicate artifacts in S3.